service/frontend/adminHandler.go

```diff
@@ -529,9 +529,87 @@ func (adh *AdminHandler) CloseShard(ctx context.Context, request *adminservice.C
+func (adh *AdminHandler) ListHistoryTasks(
+	ctx context.Context,
+	request *adminservice.ListHistoryTasksRequest,
+) (_ *adminservice.ListHistoryTasksResponse, retError error) {
+	defer log.CapturePanic(adh.logger, &retError)
+
+	scope, sw := adh.startRequestProfile(metrics.AdminListHistoryTasksScope)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	taskRange := request.GetTaskRange()
+	if taskRange == nil {
+		return nil, adh.error(errTaskRangeNotSet, scope)
+	}
+
+	var taskCategory tasks.Category
+	switch request.Category {
+	case enumsspb.TASK_CATEGORY_UNSPECIFIED:
+		return nil, adh.error(errTaskCategoryNotSet, scope)
+	case enumsspb.TASK_CATEGORY_TRANSFER:
+		taskCategory = tasks.CategoryTransfer
+	case enumsspb.TASK_CATEGORY_TIMER:
+		taskCategory = tasks.CategoryTimer
+	case enumsspb.TASK_CATEGORY_REPLICATION:
+		taskCategory = tasks.CategoryReplication
+	case enumsspb.TASK_CATEGORY_VISIBILITY:
+		taskCategory = tasks.CategoryVisibility
+	default:
+		var ok bool
+		taskCategory, ok = tasks.GetCategoryByID(int32(request.Category))
+		if !ok {
+			return nil, adh.error(&serviceerror.InvalidArgument{
+				Message: fmt.Sprintf("unknown task category ID: %v", int32(request.Category)),
+			}, scope)
+		}
+	}
+
+	var minTaskKey, maxTaskKey tasks.Key
+	if taskRange.InclusiveMinTaskKey != nil {
+		minTaskKey.FireTime = timestamp.TimeValue(taskRange.InclusiveMinTaskKey.FireTime)
+		minTaskKey.TaskID = taskRange.InclusiveMinTaskKey.TaskId
+	}
+	if taskRange.ExclusiveMaxTaskKey != nil {
+		maxTaskKey.FireTime = timestamp.TimeValue(taskRange.ExclusiveMaxTaskKey.FireTime)
+		maxTaskKey.TaskID = taskRange.ExclusiveMaxTaskKey.TaskId
+	}
+	resp, err := adh.persistenceExecutionManager.GetHistoryTasks(&persistence.GetHistoryTasksRequest{
+		ShardID:             request.ShardId,
+		TaskCategory:        taskCategory,
+		InclusiveMinTaskKey: minTaskKey,
+		ExclusiveMaxTaskKey: maxTaskKey,
+		BatchSize:           int(request.BatchSize),
+		NextPageToken:       request.NextPageToken,
+	})
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	return &adminservice.ListHistoryTasksResponse{
+		Tasks:         toAdminTask(resp.Tasks),
+		NextPageToken: resp.NextPageToken,
+	}, nil
+}
+
+func toAdminTask(tasks []tasks.Task) []*adminservice.Task {
+	var adminTasks []*adminservice.Task
+	for _, task := range tasks {
+		adminTasks = append(adminTasks, &adminservice.Task{
+			NamespaceId: task.GetNamespaceID(),
+			WorkflowId:  task.GetWorkflowID(),
+			RunId:       task.GetRunID(),
+			TaskId:      task.GetTaskID(),
+			TaskType:    task.GetType(),
+			FireTime:    timestamp.TimePtr(task.GetKey().FireTime),
+			Version:     task.GetVersion(),
+		})
+	}
+	return adminTasks
+}
 
 // ListTimerTasks lists timer tasks for a given shard
 func (adh *AdminHandler) ListTimerTasks(ctx context.Context, request *adminservice.ListTimerTasksRequest) (_ *adminservice.ListTimerTasksResponse, retError error) {
```

Comment:
Should those be marked as deprecated and be removed after some time? (After 1.17 released?)

Reply:
yeah, let me add a TODO for that.

Reply:
https://github.com/temporalio/temporal/issues/2587

Reply:
Since this is internal API and currently the TCTL command using this API is broken. We don't have to follow the normal deprecation procedure. I will send out a separate diff for the deletion, don't want to mix adding and deleting APIs. 

For this PR, I marked the rpc and request/response message as deprecated.